### ISSUE TITLE
Do we really need `publicKey` to be outside of the `request` property?

### ISSUE THREAD

**Issue by samuelgoto** [WICG/digital-identities#49]
In [this](https://github.com/WICG/identity-credential/blob/main/digital-credentials-2-proposal.md), `publicKey` is part of the top-level parameter. 

```javascript
dictionary IdentityRequestProvider {
 required DOMString protocol; // See "Protocol Registry" below
 required DOMString request;
 required DOMString publicKey;
}
```

Should `publicKey` be (opaquely, optionally and per protocol) part of `request` instead of `IdentityRequestProvider`?

**Comment by hlozi** [comment]
So, having the public key on the top level is part of making encryption work for all protocols. It's like having a common response encryption rule that everyone using the Web API follows. This has a lot of benefits for security. For one, it makes encryption more consistent and easier to implement, which means there's less chance of mistakes or weaknesses.
Response encryption provides security mitigations like phishing protection. This requires interaction between the browser and the wallet application. Standardizing this makes it easier for browsers and wallets to interact.
So, in a nutshell, using a standardized HPKE Encryption for response encryption when using the Web API, makes encryption work across different protocols while the wallet still does the encryption by itself of the response E2E to the RP.

**Comment by OR13** [comment]
`DOMString publicKey` .. I'd prefer a more precise type for the public key, recommend jwk mandatory to support, pem optional, etc... don't like the idea of arbitrary strings.

**Comment by hlozi** [comment]
It could also be a certificate. But agreed that the field "publicKey " should be well-defined.

**Comment by timcappalli** [comment]
2023-12-28 call:

Consensus on the following:

- the public key property should be required and should also live at the top level (independent of protocol)
- the value should be more structured than just a DOMString
- may be redundant with a key acquired via an out of band method, but cannot be contradictory
- need to further refine HPKE, including crypto agility

**Comment by marcoscaceres** [comment]
Could publicKey be a [`CryptoKey`](https://www.w3.org/TR/WebCryptoAPI/#dfn-CryptoKey)?

```idl
interface CryptoKey {
  readonly attribute KeyType type;
  readonly attribute boolean extractable;
  readonly attribute object algorithm;
  readonly attribute object usages;
};
```

**Comment by marcoscaceres** [comment]
Moved the discussion around the shape of the public key here: https://github.com/WICG/identity-credential/issues/56

However, I'd still like to discuss where `.publicKey` actually sits. Should it sit at the `IdentityRequestProvider` level or all the way up at the `IdentityRequestOptions` dictionary? 

I.e., put differently, will each `IdentityRequestProvider` need a different public key or can they all use the same one?

**Comment by OR13** [comment]
It seems like the most defensive thing would be to allow each provider to be initialized with a different public key or certificate.

Are there cases where the same provider might want to initialize instances with different crypto?

For example a single provider supporting X-Wing (kyber+x25519) and also supporting DHKem with P-384?

**Comment by marcoscaceres** [comment]
Yeah, you are right @OR13. I wanted to double check I hadn't missed something. 

@samuelgoto, ok to close this?

**Comment by jogu** [comment]
Can someone please explain the rational for pulling the encryption key up into this level?

I read the rational above but it makes no sense to me. I believe we're talking about the key that the verifier provides to the wallet, that the wallet will use to encrypt the credential in the response. The browser is not at all involved in the encryption or decryption process and I cannot see how having a key at the top level helps the encryption work better as the encryption is done by the wallet and exactly what is encrypted is protocol dependent.

**Comment by marcoscaceres** [comment]
>Can someone please explain the rational for pulling the encryption key up into this level?  

Structured primitive allows us to hook into browser's key processing machinery that already exists (saving us a lot of code and giving us security assurances), and allows us to easily reject certificates of the wrong type (e.g., passing a private key, or an unsupported format). 


> The browser is not at all involved in the encryption or decryption process and I cannot see how having a key at the top level helps the encryption work better as the encryption is done by the wallet and exactly what is encrypted is protocol dependent.

Browser is the first line of defense and can filter out invalid or malicious keys. We definitely don't want to be passing up arbitrary data up to wallets or the OS, as that opens a significant attack vector.

**Comment by jogu** [comment]
Thanks @marcoscaceres, that makes more sense.

I think my concern would be this may not really be applicable in some cases. Some ecosystems will define their own mechanisms for distributing keys (for example they might require that the key is looked up in a central registry based on the verifier's domain name). Would that mean there are then two keys in play and the wallet now has to deal with the case where the two keys are different?

**Comment by marcoscaceres** [comment]
> Would that mean there are then two keys in play and the wallet now has to deal with the case where the two keys are different?

I don't believe it matters on a "primitive" level: Consider, digital certificates / keys are all "crypto keys", right? That is, they are blobs that adhere to what the Web Crypto spec says they are and how they are used on the Web must be as a "crypto key", as defined by the Web Crypto API. 

To be sure, this is how [Web Crypto defines `CryptoKey`](https://www.w3.org/TR/WebCryptoAPI/#cryptokey-interface-description): 

> [A CryptoKey is] a uniform interface for many different kinds of keying material managed by the user agent. This may include keys that have been generated by the user agent, derived from other keys by the user agent, imported to the user agent through user actions or using this API, pre-provisioned within software or hardware to which the user agent has access or made available to the user agent in other ways. The term key refers broadly to any keying material including actual keys for cryptographic operations and secret values obtained within key derivation or exchange operations.

How then the underlying system works with them comes into play, but from within a web page, when we are talking JavaScript, these are `CryptoKey` objects that get passed around.

**Comment by Sakurann** [comment]
what would the verifier pass in this `publicKey` parameter if ` verifier requires the key to be looked up in a central registry based on the verifier's domain name`? Not sure it is realistic/reasonable for the browsers to obtain public keys from those registries on behalf of the wallet..?

**Comment by timcappalli** [comment]
2024-02-08 call:
- not clear if this attribute is needed at the top level as the browser is not directly using the key or interfacing with a key store
- if the attribute stays, might be better to rename as it is causing confusion

**Comment by samuelgoto** [comment]
The last spec draft does not contain publicKey at the top level, so this resolved my original concern, so resolving. Feel free to reopen if there is still something actionable here.
